### PR TITLE
[WIP] Add helper function to align labels between datasets and model config

### PR DESCRIPTION
**Pull request by lewtun**
# What does this PR do?

This PR adds a helper function to align the `label2id` and `id2label` mappings between a `datasets.Dataset` and `PretrainedConfig`, with the alignment performed on the dataset itself.

This will help us with the Hub evaluation, where we won't know in advance whether a model that is fine-tuned on say MNLI has the same mappings as the MNLI dataset we load from `datasets`.

An example where this is needed is if we naively try to evaluate `microsoft/deberta-base-mnli` on `mnli` because the model config has the following mappings:

```python
  "id2label": {
    "0": "CONTRADICTION",
    "1": "NEUTRAL",
    "2": "ENTAILMENT"
  },
  "label2id": {
    "CONTRADICTION": 0,
    "ENTAILMENT": 2,
    "NEUTRAL": 1
  }
```

while the `mnli` dataset has the `contradiction` and `neutral` labels swapped:

```python
id2label = {0: 'entailment', 1: 'neutral', 2: 'contradiction'}
label2id = {'contradiction': 2, 'entailment': 0, 'neutral': 1}
```

As a result, we get a much lower accuracy during evaluation:

```python
from datasets import load_dataset
from transformers.trainer_utils import EvalPrediction
from transformers import AutoModelForSequenceClassification, Trainer

# load dataset for evaluation
mnli = load_dataset("glue", "mnli", split="test")
# load model
model_ckpt = "microsoft/deberta-base-mnli"
model = AutoModelForSequenceClassification.from_pretrained(checkpoint)
# preprocess, create trainer ...
mnli_enc = ...
trainer = Trainer(model, args=args, tokenizer=tokenizer)
# generate preds
preds = trainer.predict(mnli_enc)
# preds.label_ids misalinged with model.config => returns wrong accuracy (too low)!
compute_metrics(EvalPrediction(preds.predictions, preds.label_ids))
```

The fix is to use the helper function before running the evaluation to make sure the label IDs are aligned:

```python
from transformers.modeling_utils import align_dataset_labels_with_config

mnli_enc_aligned = align_dataset_labels_with_config(dataset=mnli_enc, config=model.config, label_column="label")
# preds now aligned and everyone is happy :)
preds = trainer.predict(mnli_enc_aligned)
```

cc @thomwolf @lhoestq 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -1993,3 +1994,23 @@ def forward(self, hidden_states):
...
+) -> Dataset:
+    def _process_labels(batch):
+        int2str_function = dataset.features[label_column].int2str
+        label_name_to_id = {k.lower(): v for k, v in config.label2id.items()}
+        label_names = int2str_function(batch[label_column])
+        # TODO(lewtun): we need to also lowercase dataset label names?
+        batch[label_column] = label_name_to_id[label_names]
+        return batch
+
+    dataset = dataset.map(_process_labels)
```
Comment (lhoestq) [inline src/transformers/modeling_utils.py:2010]: You can pass `features=` directly in `map` instead of casting afterwards.

Also some minor optimization stuff:
- please use a batched map, this will be faster
- compute `label_name_to_id` and `int2str_function` outside the `_process_labels` function

**Comment by lewtun** [comment]
Closing this in favour of implementing generic functionality on the `datasets` side here: https://github.com/huggingface/datasets/pull/2457
